### DISCUSSION TITLE
Possibility of using sqlite update_hooks?

### DISCUSSION THREAD

**Discussion by thisguycodes** [pocketbase/pocketbase#5073] [Q&A] [answered]
This is a very high level theoretical, but hear me out:

First, I'm under the assumption that pocketbase depends on any database modifications going through it's own handlers, so that it can e.g. stream real-time updates to clients. As such if, say, a separate process were to modify the sqlite database then pocketbase wouldn't be able to detect this change.

However, sqlite itself has a [hook](https://www.sqlite.org/c3ref/update_hook.html) that can be used to function in these situations, and at least one golang sqlite library [supports using it](https://pkg.go.dev/github.com/mattn/go-sqlite3#SQLiteConn.RegisterUpdateHook).

My question is thus: could pocketbase be made to rely on this hook mechanism instead? And if so is the complexity closer to "totally rewrite the datastore pieces" or "replace a few functions in the connection stuff to instead rely on a callback".

Why?: If pocketbase were to rely on this mechanism instead, tools like [LiteFS](https://fly.io/docs/litefs/) could be used to get lower read latencies / faster streamed updates further from the "primary" version of the app (LiteFS proxies writes to the primary, but reads are served locally). There's more investment in sqlite tooling lately and it'd be awesome to be able to take advantage of them.

**Comment by ganigeorgiev** [comment, marked as answer]
This was a very early decision. Even if the pure Go driver had support for the SQLite hooks (_see https://gitlab.com/cznic/sqlite/-/issues/125_) I don't think it can be easily integrated with PocketBase because the app event hooks are unfortunately more complicated than just reacting to a db change event. 

Even the simplest db hooks (eg. `OnModel*`) require access to the transactional context or has some special handling like the `OnModelAfter*` hooks that are triggered only after a successful operation (transaction committed). With the [ongoing refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) we also differentiate and allow intercepting several stages of a "create/update/delete" operation, including modifying what is getting written to the db, data validations, before/during/after transaction, etc.

The app level event hooks are just easier to work with and allow more control.

Can the SQLite hooks and/or triggers be used - to some extend probably yes, but for me personally it is not worth currently the dev effort, especially if the only reason for this is horizontal scaling or some other distributed setup. I don't exactly understand what do you mean by "_...LiteFS could be used to get lower read latencies / faster streamed updates..._" but I can see the benefit of being able to react to changes made from external processes, but then again most PocketBase hooks require more context to operate than just the db action which means that we'll have to somehow serialize and keep track of additional data alongside the main operation. It is certainly doable but I'm not planning investing time on this at the moment. I guess to some extend this could be also implemented externally by listening the corresponding SQLite update event and then manually construct and invoke the app event [`hook.Trigger` method](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.13/tools/hook#Hook.Trigger) (_note that the hook implementation has slightly changed with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355)_).

**Reply by thisguycodes** [reply]
Thank you for the incredibly detailed response! I super appreciate all the context you gave me.
